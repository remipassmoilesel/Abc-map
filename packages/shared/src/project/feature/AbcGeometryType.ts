/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * This enum is the exact copy of openlayers/src/ol/geom/GeometryType.js because:
 * - We need our own type for serialization of projects
 * - And current typings are borked (06/04/2022), these constants are currently typed as strings
 */
export enum AbcGeometryType {
  POINT = 'Point',
  LINE_STRING = 'LineString',
  POLYGON = 'Polygon',
  LINEAR_RING = 'LinearRing',
  MULTI_POINT = 'MultiPoint',
  MULTI_LINE_STRING = 'MultiLineString',
  MULTI_POLYGON = 'MultiPolygon',
  GEOMETRY_COLLECTION = 'GeometryCollection',
  CIRCLE = 'Circle',
}

type OpenlayersGeometryType =
  | 'Point'
  | 'LineString'
  | 'LinearRing'
  | 'Polygon'
  | 'MultiPoint'
  | 'MultiLineString'
  | 'MultiPolygon'
  | 'GeometryCollection'
  | 'Circle';

export function toAbcGeometryType(type: OpenlayersGeometryType): AbcGeometryType {
  switch (type) {
    case 'Point':
      return AbcGeometryType.POINT;
    case 'LineString':
      return AbcGeometryType.LINE_STRING;
    case 'LinearRing':
      return AbcGeometryType.LINEAR_RING;
    case 'Polygon':
      return AbcGeometryType.POLYGON;
    case 'MultiPoint':
      return AbcGeometryType.MULTI_POINT;
    case 'MultiLineString':
      return AbcGeometryType.MULTI_LINE_STRING;
    case 'MultiPolygon':
      return AbcGeometryType.MULTI_POLYGON;
    case 'GeometryCollection':
      return AbcGeometryType.GEOMETRY_COLLECTION;
    case 'Circle':
      return AbcGeometryType.CIRCLE;
  }
}
