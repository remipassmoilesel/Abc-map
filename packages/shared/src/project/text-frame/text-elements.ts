/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { FeatureStyle } from '../style';
import { AbcGeometryType } from '../feature';

/**
 * This is a text slice used in text editor
 */
export type AbcText = {
  text: string;
  size?: number;
  bold?: boolean;
  italic?: boolean;
  underline?: boolean;
  foregroundColor?: string;
  backgroundColor?: string;
};

/**
 * This is a block used in text editor
 */
export type AbcElement =
  | ParagraphElement
  | CodeElement
  | QuoteElement
  | TitleElement
  | TableElement
  | TableRowElement
  | TableCellElement
  | VideoElement
  | ImageElement
  | ListElement
  | ListItemElement
  | LinkElement
  | MapSymbolElement;

export interface ParagraphElement {
  type: 'paragraph';
  align?: TextAlign;
  children: (AbcElement | AbcText)[];
}

export type TextAlign = 'left' | 'center' | 'right' | 'justify';

export interface CodeElement {
  type: 'code';
  children: AbcText[];
}

export interface QuoteElement {
  type: 'quote';
  children: AbcText[];
}

export interface TitleElement {
  type: 'title';
  level: number;
  children: AbcText[];
  align?: TextAlign;
}

export interface TableElement {
  type: 'table';
  children: TableRowElement[];
  border?: number;
}

export interface TableRowElement {
  type: 'table-row';
  children: TableCellElement[];
}

export interface TableCellElement {
  type: 'table-cell';
  // We must use paragraph inside tables, otherwise we can not have multiple lines cells
  children: ParagraphElement[];
}

export interface VideoElement {
  type: 'video';
  url: string;
  // Content is not editable
  children: [{ text: '' }];
}

export interface ImageElement {
  type: 'image';
  url: string;
  size: number;
  // Content is not editable
  children: [{ text: '' }];
}

export interface ListElement {
  type: 'list';
  ordered: boolean;
  children: ListItemElement[];
}

export interface ListItemElement {
  type: 'list-item';
  children: AbcText[];
}

export interface LinkElement {
  type: 'link';
  url: string;
  children: AbcText[];
}

export interface MapSymbolElement {
  type: 'map-symbol';
  style: FeatureStyle;
  geometryType: AbcGeometryType;
  // Content is not editable
  children: [{ text: '' }];
}
