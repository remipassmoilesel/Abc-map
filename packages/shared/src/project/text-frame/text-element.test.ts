/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcElement, AbcText } from './text-elements';
import { AbcGeometryType } from '../feature';

/**
 * If this test fail, you should write a migration script then adapt it
 */
describe('Text elements', () => {
  it('texts should not change without migration', () => {
    const actual: AbcText = {
      text: 'text1',
      size: 4,
      bold: true,
      italic: true,
      underline: true,
      foregroundColor: 'foregroundColor',
      backgroundColor: 'backgroundColor',
    };

    expect(actual).toMatchSnapshot();
  });

  it('elements should not change without migration', () => {
    const actual: AbcElement[] = [
      { type: 'paragraph', align: 'justify', children: [] },
      { type: 'paragraph', align: 'center', children: [] },
      { type: 'paragraph', align: 'left', children: [] },
      { type: 'paragraph', align: 'right', children: [] },
      { type: 'code', children: [] },
      { type: 'quote', children: [] },
      { type: 'title', level: 1, align: 'center', children: [] },
      { type: 'table', children: [] },
      { type: 'table-row', children: [] },
      { type: 'table-cell', children: [] },
      { type: 'video', url: 'http://video.url', children: [{ text: '' }] },
      { type: 'image', url: 'http://image.blob.url', size: 1, children: [{ text: '' }] },
      { type: 'list', ordered: true, children: [] },
      { type: 'list', ordered: false, children: [] },
      { type: 'list-item', children: [] },
      { type: 'link', url: 'http://somewhere.net', children: [] },
      { type: 'map-symbol', style: { fill: { color1: 'red' } }, geometryType: AbcGeometryType.POINT, children: [{ text: '' }] },
    ];

    expect(actual).toMatchSnapshot();
  });
});
