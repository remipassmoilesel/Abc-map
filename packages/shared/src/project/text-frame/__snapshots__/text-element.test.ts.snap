// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text elements elements should not change without migration 1`] = `
Array [
  Object {
    "align": "justify",
    "children": Array [],
    "type": "paragraph",
  },
  Object {
    "align": "center",
    "children": Array [],
    "type": "paragraph",
  },
  Object {
    "align": "left",
    "children": Array [],
    "type": "paragraph",
  },
  Object {
    "align": "right",
    "children": Array [],
    "type": "paragraph",
  },
  Object {
    "children": Array [],
    "type": "code",
  },
  Object {
    "children": Array [],
    "type": "quote",
  },
  Object {
    "align": "center",
    "children": Array [],
    "level": 1,
    "type": "title",
  },
  Object {
    "children": Array [],
    "type": "table",
  },
  Object {
    "children": Array [],
    "type": "table-row",
  },
  Object {
    "children": Array [],
    "type": "table-cell",
  },
  Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "type": "video",
    "url": "http://video.url",
  },
  Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "size": 1,
    "type": "image",
    "url": "http://image.blob.url",
  },
  Object {
    "children": Array [],
    "ordered": true,
    "type": "list",
  },
  Object {
    "children": Array [],
    "ordered": false,
    "type": "list",
  },
  Object {
    "children": Array [],
    "type": "list-item",
  },
  Object {
    "children": Array [],
    "type": "link",
    "url": "http://somewhere.net",
  },
  Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "geometryType": "Point",
    "style": Object {
      "fill": Object {
        "color1": "red",
      },
    },
    "type": "map-symbol",
  },
]
`;

exports[`Text elements texts should not change without migration 1`] = `
Object {
  "backgroundColor": "backgroundColor",
  "bold": true,
  "foregroundColor": "foregroundColor",
  "italic": true,
  "size": 4,
  "text": "text1",
  "underline": true,
}
`;
